core/shared/src/main/scala/zio/logging/LogFilter.scala

```diff
@@ -0,0 +1,375 @@
+   */
+  final def cacheWith[A](group: LogGroup[A]): LogFilter[Message] =
+    new LogFilter[Message] {
+      private val cache = new java.util.concurrent.ConcurrentHashMap[A, Boolean]()
+
+      override def apply(
+        trace: Trace,
+        fiberId: FiberId,
+        logLevel: LogLevel,
+        message: () => Message,
+        cause: Cause[Any],
+        context: FiberRefs,
+        spans: List[LogSpan],
+        annotations: Map[String, String]
+      ): Boolean = {
+        val key = group(trace, logLevel, context, annotations)
+        cache.computeIfAbsent(
+          key,
+          _ => self(trace, fiberId, logLevel, message, cause, context, spans, annotations)
+        )
+      }
+    }
+
+  /**
+   * Returns a version of logger that only logs messages when this filter is satisfied
+   */
+  def filter[M <: Message, O](logger: zio.ZLogger[M, O]): zio.ZLogger[M, Option[O]] =
+    new ZLogger[M, Option[O]] {
+      override def apply(
+        trace: Trace,
+        fiberId: FiberId,
+        logLevel: LogLevel,
+        message: () => M,
+        cause: Cause[Any],
+        context: FiberRefs,
+        spans: List[LogSpan],
+        annotations: Map[String, String]
+      ): Option[O] =
+        if (self(trace, fiberId, logLevel, message, cause, context, spans, annotations)) {
+          Some(logger(trace, fiberId, logLevel, message, cause, context, spans, annotations))
+        } else None
+    }
+}
+
+object LogFilter {
+
+  /**
+   * Log filter which accept all logs (logs are not filtered)
+   */
+  val acceptAll: LogFilter[Any] =
+    (
+      _: Trace,
+      _: FiberId,
+      _: LogLevel,
+      _: () => Any,
+      _: Cause[Any],
+      _: FiberRefs,
+      _: List[LogSpan],
+      _: Map[String, String]
+    ) => true
+
+  /**
+   * Returns a filter which accept logs when the log level satisfies the specified predicate
+   */
+  def logLevel(predicate: LogLevel => Boolean): LogFilter[Any] =
+    (
+      _: Trace,
+      _: FiberId,
+      logLevel: LogLevel,
+      _: () => Any,
+      _: Cause[Any],
+      _: FiberRefs,
+      _: List[LogSpan],
+      _: Map[String, String]
+    ) => predicate(logLevel)
+
+  /**
+   * Returns a filter which accept logs when the log level priority is higher then given one
+   */
+  def logLevel(rootLevel: LogLevel): LogFilter[Any] =
+    logLevel(_ >= rootLevel)
+
+  /**
+   * Defines a filter for log filtering based log level specified by given groups,
+   *
+   * filter will use log level from first matching grouping or root level, if specific log level is not found
+   *
+   * @param rootLevel Default log level
+   * @param group Log group
+   * @param matcher Mather for log group and groupings
+   * @param groupings Log levels definitions
+   * @return A filter for log filtering based on given groups
+   */
+  def logLevelByGroup[A](
+    rootLevel: LogLevel,
+    group: LogGroup[A],
+    matcher: (A, A) => Boolean,
```

Comment:
I am not sure, but could it make sense to push the `(A, A) => Boolean` into `LogGroup` as a sort of `isEquivalent`? By default it could be `l == r`, but then you could override that to have specific logic for equivalence of two `A` groups.

In fact, maybe we should have: `final case class LogGroupEquivalence[A](equivalent: (A, A) => Boolean)` just so we have a place to put constructors. Then we could consider putting that into `LogGroup` too.